File: rust/luhn.rs
/// Check a Luhn checksum.
pub fn is_valid(code: &str) -> bool {
    let code = code.replace(" ", "");
    if code.len() <= 1 {
        return false;
    }

    let mut is_even = false;
    let mut sum = 0;
    for c in code.chars().rev() {
        if !c.is_digit(10) {
            return false;
        }
        let mut n = c.to_digit(10).unwrap();
        if is_even {
            n = n * 2;
            if n >= 10 {
                n = n - 9;
            }
        }
        sum += n;
        is_even = !is_even;
    }
    sum % 10 == 0
}

File: zig/luhn.zig
const std = @import("std");

pub fn isValid(s: []const u8) bool {
    if (s.len < 2) {
        return false;
    }
    var sum: u32 = 0;
    var i: usize = s.len;
    var p: u32 = 0;
    while (i > 0) : (i -= 1) {
        const c = s[i - 1];
        if (c == ' ') {
            continue;
        }
        if (!std.ascii.isDigit(c)) {
            return false;
        }
        var n: u8 = @intCast(c - '0');
        if (p % 2 == 1) {
            n *= 2;
            if (n > 9) {
                n -= 9;
            }
        }
        sum += n;
        p += 1;
    }
    return p > 1 and sum % 10 == 0;
}
